src/index/index.c

```diff
@@ -40,79 +26,47 @@ int compareNumerics(SIValue *a, SIValue *b) {
 /* The index must maintain its own copy of the indexed SIValue
  * so that it becomes outdated but not broken by updates to the property. */
 SIValue* cloneKey(SIValue *property) {
-  SIValue *clone = malloc(sizeof(SIValue));
+  SIValue *clone = rm_malloc(sizeof(SIValue));
   *clone = SI_Clone(*property);
   return clone;
 }
 
 void freeKey(SIValue *key) {
   SIValue_Free(key);
-  free(key);
-}
-
-/* Construct key and retrieve index from Redis keyspace */
-Index* Index_Get(RedisModuleCtx *ctx, const char *graph, const char *label, const char *property) {
-  Index *idx = NULL;
-  // Open key with read-only access
-  RedisModuleKey *key = _index_LookupKey(ctx, graph, label, property, false);
-
-  if (RedisModule_ModuleTypeGetType(key) == IndexRedisModuleType) {
-    idx = RedisModule_ModuleTypeGetValue(key);
-  }
-
-  RedisModule_CloseKey(key);
-
-  return idx;
-}
-
-int Index_Delete(RedisModuleCtx *ctx, const char *graphName, const char *label, const char *prop) {
-  // Open key with write access
-  RedisModuleKey *key = _index_LookupKey(ctx, graphName, label, prop, true);
-  if (RedisModule_ModuleTypeGetType(key) != IndexRedisModuleType) {
-    // Reply with error if this key does not exist or does not correspond to an index object
-    RedisModule_CloseKey(key);
-    return INDEX_FAIL;
-  }
-
-  Index *idx = RedisModule_ModuleTypeGetValue(key);
-
-  // The DeleteKey routine will free the index and skiplists
-  RedisModule_DeleteKey(key);
-  return INDEX_OK;
+  rm_free(key);
 }
 
 void initializeSkiplists(Index *index) {
   index->string_sl = skiplistCreate(compareStrings, compareNodes, cloneKey, freeKey);
   index->numeric_sl = skiplistCreate(compareNumerics, compareNodes, cloneKey, freeKey);
 }
 
-/* buildIndex allocates an Index object and populates it with a label-property pair
- * by traversing a label matrix with a TuplesIter. */
-Index* buildIndex(Graph *g, const GrB_Matrix label_matrix, const char *label, const char *prop_str) {
-  Index *index = malloc(sizeof(Index));
+/* Index_Create allocates an Index object and populates it with a label-property pair
+ * by accessing DataBlock elements referred to by a TuplesIter over a label matrix. */
+Index* Index_Create(DataBlock *entities, TuplesIter *it, const char *label, const char *prop_str) {
+  Index *index = rm_malloc(sizeof(Index));
 
-  index->label = strdup(label);
-  index->property = strdup(prop_str);
+  index->label = rm_strdup(label);
+  index->property = rm_strdup(prop_str);
 
   initializeSkiplists(index);
 
-  TuplesIter *it = TuplesIter_new(label_matrix);
-
-  Node *node;
   EntityProperty *prop;
 
   skiplist *sl;
-  NodeID node_id;
+  EntityID entity_id;
+  GraphEntity *entity;
+
   int found;
   int prop_index = 0;
-  while(TuplesIter_next(it, NULL, &node_id) != TuplesIter_DEPLETED) {
-    node = Graph_GetNode(g, node_id);
+  while(TuplesIter_next(it, NULL, &entity_id) != TuplesIter_DEPLETED) {
+    entity = DataBlock_GetItem(entities, entity_id);
```

Comment:
I think we should get an entity from the graph and not directly from a pointer to the internal datablock.